### PR TITLE
Add Context API support

### PR DESCRIPTION
**Pull request by tmhrbr**

### CODE REVIEW

**Comment by kLabz** [comment]
What version of the context API are you using?

**Comment by tmhrbr** [comment]
> What version of the context API are you using?

Latest one that is available since [16.6.](https://5c11762d4be4d10008916ab1--reactjs.netlify.com/docs/context.html#classcontexttype)

**Comment by kLabz** [comment]
I don't think it's supposed to be used with ReactComponent's `context` field

Edit: it can, but is typed via the static field `contextType`

**Comment by tmhrbr** [comment]
> I don't think it's supposed to be used with ReactComponent's `context` field

Take a look here please https://reactjs.org/docs/context.html#classcontexttype. Based on the doc I believe it is. I can create a PoC in the sample if you want me to.

**Comment by kLabz** [comment]
It's said on your link that you need `contextType` to provide the actual context instance to populate `this.context`.

I'm not sure adding yet another type parameter to ReactComponent is necessary. We could remove the field from ReactComponent and add it to the components actually using it, with the expected typing (+ `contextType`).

**Comment by tmhrbr** [comment]
Removing `this.context` would work as well. I can hide it behind a flag so it's compatible with [this PR](https://github.com/massiveinteractive/haxe-react/pull/108/files#diff-234edfbe7a6a6b6519760b87dcf256eaL16).

**Comment by kLabz** [comment]
It would be great since it has been "deprecated" in react-next too with this define :) Working on an update to context API there, btw.

**Comment by tmhrbr** [comment]
> It would be great since it has been "deprecated" in react-next too with this define :) Working on an update to context API there, btw.

Great, I updated the PR, please take a look.

File: src/lib/react/React.hx [hunk 1]
```diff
@@ -28,6 +28,12 @@ extern class React
 	**/
 	public static function isValidElement(object:Dynamic):Bool;
 
+	/**
+		https://reactjs.org/docs/context.html#reactcreatecontext
+		Note: this API has been introduced in React 16.3
+	**/
+	public static function createContext<T>(?defaultValue:T):ReactContext<T>;
```
Comment (kLabz) [inline src/lib/react/React.hx]: `createContext` also accepts an optional second argument, `calculateChangedBits` of type `T->T->Int`.

https://github.com/facebook/react/blob/master/packages/react/src/ReactContext.js#L19
(I know, it's not documented)

File: src/lib/react/React.hx [hunk 2]
```diff
@@ -87,3 +93,9 @@ extern interface ReactChildren
 }
 
 typedef CreateElementType = haxe.extern.EitherType<haxe.extern.EitherType<String, haxe.Constraints.Function>, Class<ReactComponent>>;
+
+typedef ReactContext<T> = {
+	var displayName:String;
```
Comment (kLabz) [inline src/lib/react/React.hx:98]: Is `displayName` on `ReactContext` a thing? I don't see it in sources

File: src/lib/react/React.hx [hunk 3]
```diff
@@ -87,3 +93,9 @@ extern interface ReactChildren
...
+	var Provider:Dynamic->ReactElement;
```
Comment (kLabz) [inline src/lib/react/React.hx]: More like `{value:T}->ReactElement`, though it doesn't really matter for jsx on haxe-react.

File: src/lib/react/React.hx (hunk 2 above)
Comment (tmhrbr) [reply src/lib/react/React.hx:98]: According to this https://reactjs.org/docs/context.html#contextdisplayname.

Comment (kLabz) [reply src/lib/react/React.hx:98]: I see, they're annoying with all this reflection :D Pretty much anything can get a `displayName` attribute, I guess.

**Review by kLabz — APPROVED** [review]

**Pull request merged**